### ISSUE TITLE
Crashes when attempting to display a task with a given content

### ISSUE THREAD

**Issue by axelsimon** [dmfs/opentasks#286]
Opentasks will crash everytime it reaches a certain task in a list of tasks, if the task starts with the following:

```
[]
[]
```

Basically, something very similar to the format used for checklists `[ ]` but without the space between the square brackets.
Adding the space back in makes the task readable by OpenTasks again.

Hope this helps.

**Comment by dmfs** [comment]
@lemonboston please take care of this.

**Comment by lemonboston** [comment]
@dmfs 
I could reproduce this on the Edit screen, writing the below in the Description field and pressing save (or unfocusing the field):
```
[]
[]
```
Stacktrace:
```
java.lang.StringIndexOutOfBoundsException: length=5; index=5
    at java.lang.String.charAt(Native Method)
    at org.dmfs.tasks.model.adapters.DescriptionStringFieldAdapter.extractDescription(DescriptionStringFieldAdapter.java:114)
    at org.dmfs.tasks.model.adapters.DescriptionStringFieldAdapter.get(DescriptionStringFieldAdapter.java:62)
    at org.dmfs.tasks.widget.TextFieldEditor.onContentChanged(TextFieldEditor.java:134)
    at org.dmfs.tasks.model.ContentSet.notifyUpdateListeners(ContentSet.java:601)
    at org.dmfs.tasks.model.ContentSet.put(ContentSet.java:438)
    at org.dmfs.tasks.model.adapters.StringFieldAdapter.set(StringFieldAdapter.java:112)
    at org.dmfs.tasks.model.adapters.DescriptionStringFieldAdapter.set(DescriptionStringFieldAdapter.java:88)
    at org.dmfs.tasks.widget.TextFieldEditor.updateValues(TextFieldEditor.java:123)
    at org.dmfs.tasks.widget.BaseTaskView.updateValues(BaseTaskView.java:113)
    at org.dmfs.tasks.widget.BaseTaskView.updateValues(BaseTaskView.java:104)
    at org.dmfs.tasks.EditTaskFragment.saveAndExit(EditTaskFragment.java:726)
    at org.dmfs.tasks.EditTaskFragment.onOptionsItemSelected(EditTaskFragment.java:520)
    at android.support.v4.app.Fragment.performOptionsItemSelected(Fragment.java:2070)
    at android.support.v4.app.FragmentManagerImpl.dispatchOptionsItemSelected(FragmentManager.java:2159)
    at android.support.v4.app.FragmentController.dispatchOptionsItemSelected(FragmentController.java:295)
    at android.support.v4.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:367)
    at android.support.v7.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:147)
    at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:100)
    at android.support.v7.app.AppCompatDelegateImplV7.onMenuItemSelected(AppCompatDelegateImplV7.java:620)
```

The bug is in `DescriptionStringFieldAdapter#extractDescription`, not handling this case properly. I can fix that and create some unit test cases as well, but I looked around in code a bit about how this combined description and checklist is handled and I would like to ask a few questions first, also to know how my change could affect other flows as well, or any other related code needs update, too.

1. Why are description and checklist serialized together? Is it because checklist is not part of the standard (btw, is a Task an iCalendar [VTODO](https://tools.ietf.org/html/rfc5545#page-56)?), so it can only be 'hacked into' the description field, and this is how OpenTasks and other task apps, providers handle it? Is there any non-official description about this format (whitespaces, etc)?
If they could also be saved separately on our side, is that something worth considering now?
With this combined handling, we are basically disallowing the checklist markers ([ ], [x], [X]) in description, is that okay?
2. When fixing this bug, how should we treat `[]`? Same as `[ ]` or treat it as plain text, not representing checklist item?
3. Is there any program flow that I should be aware when touching this? I mean I suppose this is used for synced tasks as well, so a task coming from a provider. What if it is coming with a format of having an extra empty line between description and checklists, but we trim that (should we?) and save and sync it back without it, then maybe the provider can't handle it.
4. I thought a bit about how description and checklist could be represented instead of those static methods for parsing, extracting, serializing, and I think the declarative approach using CharSequences may look something like this:
```
interface CheckList extends CharSequence {
	List<CheckListItem> items();
	String toString();
}

interface CheckListItem extends CharSequence {
	boolean checked();
	CharSequence itemText();
	String toString();
}

interface Description extends CharSequence {
	String toString();
}

interface DescriptionAndChecklist extends CharSequence {
	Description description();
	CheckList checkList();
	String toString();
}
```
`toString()`s would be used to serialize, and the implementation constructors to parse (maybe lazily).

Is this something we should work on now to maybe make these parts of the code more 'correct'. If it is done with unit tests as well, maybe we would lessen the chance of encountering issues with it later. However, it is already in production for long, so the chance is quite small, I suppose.
Or should I just fix the bug in the mentioned method (with some small refactoring) and add unit test to fix this bug quickly?

**Comment by dmfs** [comment]
For now just fix the bug by also allowing `[]` as an unchecked item. We don't have the time to refactor this code now. I don't expect any side effects. The current code only accepts `[ ]` sequences at the start of a line, so in general this is what you actually want.

I'm not aware of any standard for this, but some Markdown parsers support this notation to create checklists. So consider this as some "Markdown"-lite support, which also explains why it goes to the description field. Ideally we would just show the list items inline in the text, but that requires a lot more changes.

The declarative proposal has some serious flaws, IMO. A CheckList is not a `CharSequence` and should not extend this interface. If anything it should extend or implement a `List` interface. I would argue that a `CheckListItem` is not a `CharSequence` either.

**Comment by lemonboston** [comment]
@dmfs 
I found out that the bug that caused the crash have already been fixed by you on staging in commit ba6327a70a9adb317ef30e589b6c4c369f2eeb3e. I used _master_ version when reproduced it.

But I've already updated the code to handle `[]` as valid unchecked marker as well, so I am going to create a pull request, so you can check if you want to merge it. I've also added tests for this `extractDescription()` method and a javadoc.

About the proposed declarative approach above, yeah, you are right it wasn't good. Just because we need a serialized string representation of the object it doesn't mean it should be `CharSequence`, a special interface would be better (I would now argue against simply using toString() for that). And also, Checklist could be better as a List, I agree.

Thanks for comments and explaining the Markdown background of this format.
